File: javascript/pascals-triangle/pascals-triangle.js
export class Triangle {
  constructor(count) {
    this.lastRow = [1]
    this.rows = [[1]]

    for (let r = 1; r < count; r++) {
      this.lastRow = nextRow(this.rows[this.rows.length - 1])
      this.rows.push(this.lastRow)
    }
  }
}

const nextRow = (last) => {
  let row = [1]
  for (let w = 0; w + 1 < last.length; w++) {
    row.push(last[w] + last[w + 1])
  }
  row.push(1)
  return row
}

File: cpp/pascals-triangle/pascals_triangle.cpp
#include "pascals_triangle.h"
#include <vector>

using namespace std;

vector<int> row(const vector<int> &last_row) {
    vector<int> v;

    if (last_row.size() == 0) {
        v.push_back(1);
    } else {
        v.push_back(1);
        if (last_row.size() > 1) {
            for (auto itr = last_row.begin(); itr != last_row.end() - 1;
                 itr++) {
                v.push_back(*itr + *(itr + 1));
            }
        }

        v.push_back(1);
    }

    return v;
}

vector<vector<int>> pascals_triangle::generate_rows(int row_count) {
    vector<vector<int>> triangle;

    if (row_count > 0) {
        vector<int> blank;
        triangle.emplace_back(row(blank));

        for (int i = 1; i < row_count; i++) {
            triangle.emplace_back(row(triangle[i - 1]));
        }
    }

    return triangle;
}
